File: frontend/src/services/darshanService.js
import api from './api';

export const darshanService = {
    getSlotsByTemple: async (templeId, fromDate, toDate) => {
        const params = new URLSearchParams({ templeId: templeId.toString() });
        if (fromDate) params.append('fromDate', fromDate);
        if (toDate) params.append('toDate', toDate);

        const response = await api.get(`/darshan/temple/${templeId}?${params.toString()}`);
        return response.data;
    },

    getAvailableSlots: async (templeId, date) => {
        const params = new URLSearchParams({
            templeId: templeId.toString(),
            date: date,
        });

        const response = await api.get(`/darshan/available?${params.toString()}`);
        return response.data;
    },

    getById: async (id) => {
        const response = await api.get(`/darshan/${id}`);
        return response.data;
    },

    createSlot: async (slotData) => {
        try {
            console.log('Sending slot data to API:', slotData);
            const response = await api.post('/darshan', slotData);
            return response.data;
        } catch (error) {
            console.error('API Error:', {
                status: error.response?.status,
                data: error.response?.data,
                message: error.message
            });
            throw error;
        }
    },
};

File: backend-spring/src/main/java/com/tirthseva/api/service/DarshanService.java
package com.tirthseva.api.service;

import com.tirthseva.api.dto.darshan.DarshanSlotCreateRequest;
import com.tirthseva.api.dto.darshan.DarshanSlotDTO;
import com.tirthseva.api.entity.DarshanSlot;
import com.tirthseva.api.repository.DarshanSlotRepository;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.time.LocalDate;
import java.time.LocalDateTime;
import java.time.LocalTime;
import java.time.format.DateTimeFormatter;
import java.util.List;

@Service
@RequiredArgsConstructor
public class DarshanService {

    private final DarshanSlotRepository darshanSlotRepository;
    private static final DateTimeFormatter TIME_FORMATTER = DateTimeFormatter.ofPattern("HH:mm:ss");

    public List<DarshanSlot> getSlotsByTemple(Integer templeId, LocalDate fromDate, LocalDate toDate) {
        if (fromDate != null && toDate != null) {
            return darshanSlotRepository.findByTempleIdAndDateBetween(templeId, fromDate, toDate);
        }
        return darshanSlotRepository.findByTempleIdOrderByDateAscStartTimeAsc(templeId);
    }

    public List<DarshanSlot> getAvailableSlots(Integer templeId, LocalDate date) {
        return darshanSlotRepository.findByTempleIdAndDateAndAvailableSlotsGreaterThanOrderByStartTimeAsc(
                templeId, date, 0);
    }

    public DarshanSlot getSlotById(Integer id) {
        return darshanSlotRepository.findById(id).orElse(null);
    }

    @Transactional
    public DarshanSlot createSlot(DarshanSlotCreateRequest request) {
        try {
            DarshanSlot slot = DarshanSlot.builder()
                    .templeId(request.getTempleId())
                    .date(request.getDate())
                    .startTime(LocalTime.parse(request.getStartTime(), TIME_FORMATTER))
                    .endTime(LocalTime.parse(request.getEndTime(), TIME_FORMATTER))
                    .capacity(request.getCapacity())
                    .availableSlots(request.getCapacity())
                    .price(request.getPrice())
                    .createdAt(LocalDateTime.now())
                    .build();

            return darshanSlotRepository.save(slot);
        } catch (Exception e) {
            throw new RuntimeException("Failed to create darshan slot: " + e.getMessage(), e);
        }
    }

    public DarshanSlotDTO mapToDTO(DarshanSlot s) {
        return DarshanSlotDTO.builder()
                .id(s.getId())
                .templeId(s.getTempleId())
                .date(s.getDate())
                .startTime(s.getStartTime().format(DateTimeFormatter.ofPattern("HH:mm:ss")))
                .endTime(s.getEndTime().format(DateTimeFormatter.ofPattern("HH:mm:ss")))
                .capacity(s.getCapacity())
                .availableSlots(s.getAvailableSlots())
                .price(s.getPrice())
                .build();
    }
}
